app/lib/shared/name_tracker.dart

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2018, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+import 'package:gcloud/db.dart';
+
+import 'package:pub_dartlang_org/frontend/models.dart';
+
+final Logger _logger = new Logger('pub.name_tracker');
+const Duration _pollingInterval = const Duration(seconds: 10);
+
+final NameTracker nameTracker = new NameTracker();
+
+class NameTracker {
+  final Set<String> _names = new Set<String>();
+  final Set<String> _reducedNames = new Set<String>();
+
+  void add(String name) {
+    _names.add(name);
+    _reducedNames.add(_reduce(name));
+  }
+
+  bool hasPackage(String name) => _names.contains(name);
+
+  bool hasConflict(String name) => _reducedNames.contains(_reduce(name));
+
+  bool accept(String name) => hasPackage(name) || !hasConflict(name);
+
+  String _reduce(String name) => name.replaceAll('_', '').toLowerCase();
```

Comment:
We only remove `_` because we already don't allow other punctuation, right?

Reply:
Correct, added a comment for it.